// texvcjs/lib/texutil.js
// Information about TeX functions.
// In its own module so that the sets aren't recreated from scratch
// every time that parse() is called.
"use strict";

// track all known function names, so we can give good errors for unknown
// functions.
var all_functions = module.exports.all_functions = Object.create(null);
all_functions['\\begin'] = all_functions['\\end'] = true;

var arr2set = function(a) {
    // note that the fact that all keys in the set are prefixed with '\\'
    // helps avoid accidental name conflicts.  But use Object.create(null)
    // to be extra safe.
    var set = Object.create(null);
    a.forEach(function(v) {
        console.assert(!set[v], v);
        set[v] = all_functions[v] = true;
    });
    return set;
};
var obj2map = function(o) {
    // this just recreates the argument, but with `null` as prototype.
    var map = Object.create(null);
    Object.keys(o).forEach(function(f) {
        console.assert(!map[f]);
        map[f] = o[f]; all_functions[f] = true;
    });
    return map;
};

// Sets of function names
module.exports.box_functions = arr2set([
    "\\text", "\\mbox", "\\hbox", "\\vbox"
]);

module.exports.latex_function_names = arr2set([
    "\\arccos", "\\arcsin", "\\arctan", "\\arg", "\\cosh", "\\cos",
    "\\cot", "\\coth", "\\csc", "\\deg", "\\det", "\\dim", "\\exp",
    "\\gcd", "\\hom", "\\inf", "\\ker", "\\lg", "\\lim", "\\liminf",
    "\\limsup", "\\ln", "\\log", "\\max", "\\min", "\\Pr", "\\sec",
    "\\sin", "\\sinh", "\\sup", "\\tan", "\\tanh"
]);

module.exports.mediawiki_function_names = arr2set([
    "\\arccot", "\\arcsec", "\\arccsc", "\\sgn", "\\sen"
]);

module.exports.other_literals1 = arr2set([
    "\\aleph",
    "\\alpha",
    "\\amalg",
    "\\And",
    "\\angle",
    "\\approx",
    "\\approxeq",
    "\\ast",
    "\\asymp",
    "\\backepsilon",
    "\\backprime",
    "\\backsim",
    "\\backsimeq",
    "\\barwedge",
    "\\Bbbk",
    "\\because",
    "\\beta",
    "\\beth",
    "\\between",
    "\\bigcap",
    "\\bigcirc",
    "\\bigcup",
    "\\bigodot",
    "\\bigoplus",
    "\\bigotimes",
    "\\bigsqcup",
    "\\bigstar",
    "\\bigtriangledown",
    "\\bigtriangleup",
    "\\biguplus",
    "\\bigvee",
    "\\bigwedge",
    "\\blacklozenge",
    "\\blacksquare",
    "\\blacktriangle",
    "\\blacktriangledown",
    "\\blacktriangleleft",
    "\\blacktriangleright",
    "\\bot",
    "\\bowtie",
    "\\Box",
    "\\boxdot",
    "\\boxminus",
    "\\boxplus",
    "\\boxtimes",
    "\\bullet",
    "\\bumpeq",
    "\\Bumpeq",
    "\\cap",
    "\\Cap",
    "\\cdot",
    "\\cdots",
    "\\centerdot",
    "\\checkmark",
    "\\chi",
    "\\circ",
    "\\circeq",
    "\\circlearrowleft",
    "\\circlearrowright",
    "\\circledast",
    "\\circledcirc",
    "\\circleddash",
    "\\circledS",
    "\\clubsuit",
    "\\colon",
    "\\complement",
    "\\cong",
    "\\coprod",
    "\\cup",
    "\\Cup",
    "\\curlyeqprec",
    "\\curlyeqsucc",
    "\\curlyvee",
    "\\curlywedge",
    "\\curvearrowleft",
    "\\curvearrowright",
    "\\dagger",
    "\\daleth",
    "\\dashv",
    "\\ddagger",
    "\\ddots",
    "\\delta",
    "\\Delta",
    "\\diagdown",
    "\\diagup",
    "\\diamond",
    "\\Diamond",
    "\\diamondsuit",
    "\\digamma",
    "\\displaystyle",
    "\\div",
    "\\divideontimes",
    "\\doteq",
    "\\doteqdot",
    "\\dotplus",
    "\\dots",
    "\\dotsb",
    "\\dotsc",
    "\\dotsi",
    "\\dotsm",
    "\\dotso",
    "\\doublebarwedge",
    "\\downdownarrows",
    "\\downharpoonleft",
    "\\downharpoonright",
    "\\ell",
    "\\emptyset",
    "\\epsilon",
    "\\eqcirc",
    "\\eqsim",
    "\\eqslantgtr",
    "\\eqslantless",
    "\\equiv",
    "\\eta",
    "\\eth",
    "\\exists",
    "\\fallingdotseq",
    "\\Finv",
    "\\flat",
    "\\forall",
    "\\frown",
    "\\Game",
    "\\gamma",
    "\\Gamma",
    "\\geq",
    "\\geqq",
    "\\geqslant",
    "\\gets",
    "\\gg",
    "\\ggg",
    "\\gimel",
    "\\gnapprox",
    "\\gneq",
    "\\gneqq",
    "\\gnsim",
    "\\gtrapprox",
    "\\gtrdot",
    "\\gtreqless",
    "\\gtreqqless",
    "\\gtrless",
    "\\gtrsim",
    "\\gvertneqq",
    "\\hbar",
    "\\heartsuit",
    //"\\hline", // moved to hline_function
    "\\hookleftarrow",
    "\\hookrightarrow",
    "\\hslash",
    "\\iff",
    "\\iiiint",
    "\\iiint",
    "\\iint",
    "\\Im",
    "\\imath",
    "\\implies",
    "\\in",
    "\\infty",
    "\\injlim",
    "\\int",
    "\\intercal",
    "\\iota",
    "\\jmath",
    "\\kappa",
    "\\lambda",
    "\\Lambda",
    "\\land",
    "\\ldots",
    "\\leftarrow",
    "\\Leftarrow",
    "\\leftarrowtail",
    "\\leftharpoondown",
    "\\leftharpoonup",
    "\\leftleftarrows",
    "\\leftrightarrow",
    "\\Leftrightarrow",
    "\\leftrightarrows",
    "\\leftrightharpoons",
    "\\leftrightsquigarrow",
    "\\leftthreetimes",
    "\\leq",
    "\\leqq",
    "\\leqslant",
    "\\lessapprox",
    "\\lessdot",
    "\\lesseqgtr",
    "\\lesseqqgtr",
    "\\lessgtr",
    "\\lesssim",
    "\\limits", // XXX only valid in certain contexts
    "\\ll",
    "\\Lleftarrow",
    "\\lll",
    "\\lnapprox",
    "\\lneq",
    "\\lneqq",
    "\\lnot",
    "\\lnsim",
    "\\longleftarrow",
    "\\Longleftarrow",
    "\\longleftrightarrow",
    "\\Longleftrightarrow",
    "\\longmapsto",
    "\\longrightarrow",
    "\\Longrightarrow",
    "\\looparrowleft",
    "\\looparrowright",
    "\\lor",
    "\\lozenge",
    "\\Lsh",
    "\\ltimes",
    "\\lVert",
    "\\lvertneqq",
    "\\mapsto",
    "\\measuredangle",
    "\\mho",
    "\\mid",
    "\\mod",
    "\\models",
    "\\mp",
    "\\mu",
    "\\multimap",
    "\\nabla",
    "\\natural",
    "\\ncong",
    "\\nearrow",
    "\\neg",
    "\\neq",
    "\\nexists",
    "\\ngeq",
    "\\ngeqq",
    "\\ngeqslant",
    "\\ngtr",
    "\\ni",
    "\\nleftarrow",
    "\\nLeftarrow",
    "\\nleftrightarrow",
    "\\nLeftrightarrow",
    "\\nleq",
    "\\nleqq",
    "\\nleqslant",
    "\\nless",
    "\\nmid",
    "\\nolimits", // XXX see \limits
    "\\not",
    "\\notin",
    "\\nparallel",
    "\\nprec",
    "\\npreceq",
    "\\nrightarrow",
    "\\nRightarrow",
    "\\nshortmid",
    "\\nshortparallel",
    "\\nsim",
    "\\nsubseteq",
    "\\nsubseteqq",
    "\\nsucc",
    "\\nsucceq",
    "\\nsupseteq",
    "\\nsupseteqq",
    "\\ntriangleleft",
    "\\ntrianglelefteq",
    "\\ntriangleright",
    "\\ntrianglerighteq",
    "\\nu",
    "\\nvdash",
    "\\nVdash",
    "\\nvDash",
    "\\nVDash",
    "\\nwarrow",
    "\\odot",
    "\\oint",
    "\\omega",
    "\\Omega",
    "\\ominus",
    "\\oplus",
    "\\oslash",
    "\\otimes",
    //"\\overbrace", // moved to ar1nb (grabs trailing sub/superscript)
    "\\P",
    "\\parallel",
    "\\partial",
    "\\perp",
    "\\phi",
    "\\Phi",
    "\\pi",
    "\\Pi",
    "\\pitchfork",
    "\\pm",
    "\\prec",
    "\\precapprox",
    "\\preccurlyeq",
    "\\preceq",
    "\\precnapprox",
    "\\precneqq",
    "\\precnsim",
    "\\precsim",
    "\\prime",
    "\\prod",
    "\\projlim",
    "\\propto",
    "\\psi",
    "\\Psi",
    "\\qquad",
    "\\quad",
    "\\Re",
    "\\rho",
    "\\rightarrow",
    "\\Rightarrow",
    "\\rightarrowtail",
    "\\rightharpoondown",
    "\\rightharpoonup",
    "\\rightleftarrows",
    "\\rightrightarrows",
    "\\rightsquigarrow",
    "\\rightthreetimes",
    "\\risingdotseq",
    "\\Rrightarrow",
    "\\Rsh",
    "\\rtimes",
    "\\rVert",
    "\\S",
    "\\scriptscriptstyle",
    "\\scriptstyle",
    "\\searrow",
    "\\setminus",
    "\\sharp",
    "\\shortmid",
    "\\shortparallel",
    "\\sigma",
    "\\Sigma",
    "\\sim",
    "\\simeq",
    "\\smallfrown",
    "\\smallsetminus",
    "\\smallsmile",
    "\\smile",
    "\\spadesuit",
    "\\sphericalangle",
    "\\sqcap",
    "\\sqcup",
    "\\sqsubset",
    "\\sqsubseteq",
    "\\sqsupset",
    "\\sqsupseteq",
    "\\square",
    "\\star",
    "\\subset",
    "\\Subset",
    "\\subseteq",
    "\\subseteqq",
    "\\subsetneq",
    "\\subsetneqq",
    "\\succ",
    "\\succapprox",
    "\\succcurlyeq",
    "\\succeq",
    "\\succnapprox",
    "\\succneqq",
    "\\succnsim",
    "\\succsim",
    "\\sum",
    "\\supset",
    "\\Supset",
    "\\supseteq",
    "\\supseteqq",
    "\\supsetneq",
    "\\supsetneqq",
    "\\surd",
    "\\swarrow",
    "\\tau",
    "\\textstyle",
    "\\therefore",
    "\\theta",
    "\\Theta",
    "\\thickapprox",
    "\\thicksim",
    "\\times",
    "\\to",
    "\\top",
    "\\triangle",
    "\\triangledown",
    "\\triangleleft",
    "\\trianglelefteq",
    "\\triangleq",
    "\\triangleright",
    "\\trianglerighteq",
    //"\\underbrace", // moved to ar1nb (grabs trailing sub/superscript)
    "\\upharpoonleft",
    "\\upharpoonright",
    "\\uplus",
    "\\upsilon",
    "\\Upsilon",
    "\\upuparrows",
    "\\varepsilon",
    "\\varinjlim",
    "\\varkappa",
    "\\varliminf",
    "\\varlimsup",
    "\\varnothing",
    "\\varphi",
    "\\varpi",
    "\\varprojlim",
    "\\varpropto",
    "\\varrho",
    "\\varsigma",
    "\\varsubsetneq",
    "\\varsubsetneqq",
    "\\varsupsetneq",
    "\\varsupsetneqq",
    "\\vartheta",
    "\\vartriangle",
    "\\vartriangleleft",
    "\\vartriangleright",
    "\\vdash",
    "\\Vdash",
    "\\vDash",
    "\\vdots",
    "\\vee",
    "\\veebar",
    "\\vline",
    "\\Vvdash",
    "\\wedge",
    "\\wp",
    "\\wr",
    "\\xi",
    "\\Xi",
    "\\zeta"
]);

// text-mode literals; enclose in \mbox
module.exports.other_literals2 = arr2set([
    "\\AA",
    "\\Coppa",
    "\\coppa",
    "\\Digamma",
    "\\euro",
    "\\geneuro",
    "\\geneuronarrow",
    "\\geneurowide",
    "\\Koppa",
    "\\koppa",
    "\\officialeuro",
    "\\Sampi",
    "\\sampi",
    "\\Stigma",
    "\\stigma",
    "\\textvisiblespace",
    "\\varstigma"
]);

module.exports.other_literals3 = obj2map({
    "\\C": "\\mathbb{C}",
    "\\H": "\\mathbb{H}",
    "\\N": "\\mathbb{N}",
    "\\Q": "\\mathbb{Q}",
    "\\R": "\\mathbb{R}",
    "\\Z": "\\mathbb{Z}",
    "\\alef": "\\aleph",
    "\\alefsym": "\\aleph",
    "\\Alpha": "\\mathrm{A}",
    "\\and": "\\land",
    "\\ang": "\\angle",
    "\\Beta": "\\mathrm{B}",
    "\\bull": "\\bullet",
    "\\Chi": "\\mathrm{X}",
    "\\clubs": "\\clubsuit",
    "\\cnums": "\\mathbb{C}",
    "\\Complex": "\\mathbb{C}",
    "\\Dagger": "\\ddagger",
    "\\diamonds": "\\diamondsuit",
    "\\Doteq": "\\doteqdot",
    "\\doublecap": "\\Cap",
    "\\doublecup": "\\Cup",
    "\\empty": "\\emptyset",
    "\\Epsilon": "\\mathrm{E}",
    "\\Eta": "\\mathrm{H}",
    "\\exist": "\\exists",
    "\\ge": "\\geq",
    "\\gggtr": "\\ggg",
    "\\hAar": "\\Leftrightarrow",
    "\\harr": "\\leftrightarrow",
    "\\Harr": "\\Leftrightarrow",
    "\\hearts": "\\heartsuit",
    "\\image": "\\Im",
    "\\infin": "\\infty",
    "\\Iota": "\\mathrm{I}",
    "\\isin": "\\in",
    "\\Kappa": "\\mathrm{K}",
    "\\larr": "\\leftarrow",
    "\\Larr": "\\Leftarrow",
    "\\lArr": "\\Leftarrow",
    "\\le": "\\leq",
    "\\lrarr": "\\leftrightarrow",
    "\\Lrarr": "\\Leftrightarrow",
    "\\lrArr": "\\Leftrightarrow",
    "\\Mu": "\\mathrm{M}",
    "\\natnums": "\\mathbb{N}",
    "\\ne": "\\neq",
    "\\Nu": "\\mathrm{N}",
    "\\O": "\\emptyset",
    "\\omicron": "\\mathrm{o}",
    "\\Omicron": "\\mathrm{O}",
    "\\or": "\\lor",
    "\\part": "\\partial",
    "\\plusmn": "\\pm",
    "\\rarr": "\\rightarrow",
    "\\Rarr": "\\Rightarrow",
    "\\rArr": "\\Rightarrow",
    "\\real": "\\Re",
    "\\reals": "\\mathbb{R}",
    "\\Reals": "\\mathbb{R}",
    "\\restriction": "\\upharpoonright",
    "\\Rho": "\\mathrm{P}",
    "\\sdot": "\\cdot",
    "\\sect": "\\S",
    "\\spades": "\\spadesuit",
    "\\sub": "\\subset",
    "\\sube": "\\subseteq",
    "\\supe": "\\supseteq",
    "\\Tau": "\\mathrm{T}",
    "\\thetasym": "\\vartheta",
    "\\varcoppa": "\\mbox{\\coppa}",
    "\\weierp": "\\wp",
    "\\Zeta": "\\mathrm{Z}"
});

module.exports.big_literals = arr2set([
    "\\big",
    "\\Big",
    "\\bigg",
    "\\Bigg",
    "\\biggl",
    "\\Biggl",
    "\\biggr",
    "\\Biggr",
    "\\bigl",
    "\\Bigl",
    "\\bigr",
    "\\Bigr"
]);

module.exports.other_delimiters1 = arr2set([
    "\\backslash",
    "\\downarrow",
    "\\Downarrow",
    "\\langle",
    "\\lbrace",
    "\\lceil",
    "\\lfloor",
    "\\llcorner",
    "\\lrcorner",
    "\\rangle",
    "\\rbrace",
    "\\rceil",
    "\\rfloor",
    "\\rightleftharpoons",
    "\\twoheadleftarrow",
    "\\twoheadrightarrow",
    "\\ulcorner",
    "\\uparrow",
    "\\Uparrow",
    "\\updownarrow",
    "\\Updownarrow",
    "\\urcorner",
    "\\Vert",
    "\\vert",
    "\\lbrack",
    "\\rbrack"
]);

module.exports.other_delimiters2 = obj2map({
    "\\darr": "\\downarrow",
    "\\dArr": "\\Downarrow",
    "\\Darr": "\\Downarrow",
    "\\lang": "\\langle",
    "\\rang": "\\rangle",
    "\\uarr": "\\uparrow",
    "\\uArr": "\\Uparrow",
    "\\Uarr": "\\Uparrow"
});

module.exports.fun_ar1 = arr2set([
    "\\acute",
    "\\bar",
    "\\bcancel",
    "\\bmod",
    "\\boldsymbol",
    "\\breve",
    "\\cancel",
    "\\check",
    "\\ddot",
    "\\dot",
    "\\emph",
    "\\grave",
    "\\hat",
    //"\\mathbb", // moved to fun_ar1nb
    //"\\mathbf", // moved to fun_ar1nb
    "\\mathbin",
    "\\mathcal",
    "\\mathclose",
    "\\mathfrak",
    "\\mathit",
    "\\mathop",
    "\\mathopen",
    "\\mathord",
    "\\mathpunct",
    "\\mathrel",
    //"\\mathrm", // moved to fun_ar1nb
    "\\mathsf",
    "\\mathtt",
    //"\\operatorname", // already exists in fun_ar1nb
    "\\overleftarrow",
    "\\overleftrightarrow",
    "\\overline",
    "\\overrightarrow",
    "\\pmod",
    "\\sqrt",
    "\\textbf",
    "\\textit",
    "\\textrm",
    "\\textsf",
    "\\texttt",
    "\\tilde",
    "\\underline",
    "\\vec",
    "\\widehat",
    "\\widetilde",
    "\\xcancel",
    "\\xleftarrow",
    "\\xrightarrow"
]);

module.exports.other_fun_ar1 = obj2map({
    "\\Bbb": "\\mathbb",
    "\\bold": "\\mathbf"
});

module.exports.fun_ar1nb = arr2set([
    "\\operatorname",
    "\\overbrace",
    "\\mathbb",
    "\\mathbf",
    "\\mathrm",
    "\\underbrace"
]);

module.exports.fun_ar1opt = arr2set([
    "\\sqrt", "\\xleftarrow", "\\xrightarrow"
]);

module.exports.fun_ar2 = arr2set([
    "\\binom",
    "\\cancelto",
    "\\cfrac",
    "\\dbinom",
    "\\dfrac",
    "\\frac",
    "\\overset",
    "\\stackrel",
    "\\tbinom",
    "\\tfrac",
    "\\underset"
]);

module.exports.fun_ar2nb = arr2set([
    "\\sideset"
]);

module.exports.fun_infix = arr2set([
    "\\atop",
    "\\choose",
    "\\over"
]);

module.exports.declh_function = arr2set([
    "\\rm",
    "\\it",
    "\\cal",
    "\\bf"
]);

module.exports.left_function = arr2set([ "\\left" ]);
module.exports.right_function = arr2set([ "\\right" ]);
module.exports.hline_function = arr2set([ "\\hline" ]);
module.exports.definecolor_function = arr2set([ "\\definecolor" ]);
module.exports.color_function = arr2set([ "\\color", "\\pagecolor" ]);

// ------------------------------------------------------
// Package dependencies for various allowed commands.

module.exports.ams_required = arr2set([
    "\\text",
    "\\begin{matrix}",
    "\\begin{pmatrix}",
    "\\begin{bmatrix}",
    "\\begin{Bmatrix}",
    "\\begin{vmatrix}",
    "\\begin{Vmatrix}",
    "\\begin{aligned}",
    "\\begin{alignedat}",
    "\\begin{smallmatrix}",
    "\\begin{cases}",

    "\\ulcorner",
    "\\urcorner",
    "\\llcorner",
    "\\lrcorner",
    "\\twoheadleftarrow",
    "\\twoheadrightarrow",
    "\\xleftarrow",
    "\\xrightarrow",
    //"\\angle", // in texvc, but ams not actually required
    "\\sqsupset",
    "\\sqsubset",
    //"\\sqsupseteq", // in texvc, but ams not actually required
    //"\\sqsubseteq", // in texvc, but ams not actually required
    "\\smallsetminus",
    "\\And",
    //"\\sqcap", // in texvc, but ams not actually required
    //"\\sqcup", // in texvc, but ams not actually required
    "\\implies",
    "\\mod",
    "\\Diamond",
    "\\dotsb",
    "\\dotsc",
    "\\dotsi",
    "\\dotsm",
    "\\dotso",
    "\\lVert",
    "\\rVert",
    "\\nmid",
    "\\lesssim",
    "\\ngeq",
    "\\smallsmile",
    "\\smallfrown",
    "\\nleftarrow",
    "\\nrightarrow",
    "\\trianglelefteq",
    "\\trianglerighteq",
    "\\square",
    "\\checkmark",
    "\\supsetneq",
    "\\subsetneq",
    "\\Box",
    "\\nleq",
    "\\upharpoonright",
    "\\upharpoonleft",
    "\\downharpoonright",
    "\\downharpoonleft",
    //"\\rightharpoonup", // in texvc, but ams not actually required
    //"\\rightharpoondown", // in texvc, but ams not actually required
    //"\\leftharpoonup", // in texvc, but ams not actually required
    //"\\leftharpoondown", // in texvc, but ams not actually required
    "\\nless",
    "\\Vdash",
    "\\vDash",
    "\\varkappa",
    "\\digamma",
    "\\beth",
    "\\daleth",
    "\\gimel",
    "\\complement",
    "\\eth",
    "\\hslash",
    "\\mho",
    "\\Finv",
    "\\Game",
    "\\varlimsup",
    "\\varliminf",
    "\\varinjlim",
    "\\varprojlim",
    "\\injlim",
    "\\projlim",
    "\\iint",
    "\\iiint",
    "\\iiiint",
    "\\varnothing",
    "\\overleftrightarrow",
    "\\binom",
    "\\dbinom",
    "\\tbinom",
    "\\sideset",
    "\\underset",
    "\\overset",
    "\\dfrac",
    "\\tfrac",
    "\\cfrac",
    //"\\bigl", // in texvc, but ams not actually required
    //"\\bigr", // in texvc, but ams not actually required
    //"\\Bigl", // in texvc, but ams not actually required
    //"\\Bigr", // in texvc, but ams not actually required
    //"\\biggl", // in texvc, but ams not actually required
    //"\\biggr", // in texvc, but ams not actually required
    //"\\Biggl", // in texvc, but ams not actually required
    //"\\Biggr", // in texvc, but ams not actually required
    "\\vartriangle",
    "\\triangledown",
    "\\lozenge",
    "\\circledS",
    "\\measuredangle",
    "\\nexists",
    "\\Bbbk",
    "\\backprime",
    "\\blacktriangle",
    "\\blacktriangledown",
    "\\blacksquare",
    "\\blacklozenge",
    "\\bigstar",
    "\\sphericalangle",
    "\\diagup",
    "\\diagdown",
    "\\dotplus",
    "\\Cap",
    "\\Cup",
    "\\barwedge",
    "\\veebar",
    "\\doublebarwedge",
    "\\boxminus",
    "\\boxtimes",
    "\\boxdot",
    "\\boxplus",
    "\\divideontimes",
    "\\ltimes",
    "\\rtimes",
    "\\leftthreetimes",
    "\\rightthreetimes",
    "\\curlywedge",
    "\\curlyvee",
    "\\circleddash",
    "\\circledast",
    "\\circledcirc",
    "\\centerdot",
    "\\intercal",
    "\\leqq",
    "\\leqslant",
    "\\eqslantless",
    "\\lessapprox",
    "\\approxeq",
    "\\lessdot",
    "\\lll",
    "\\lessgtr",
    "\\lesseqgtr",
    "\\lesseqqgtr",
    "\\doteqdot",
    "\\risingdotseq",
    "\\fallingdotseq",
    "\\backsim",
    "\\backsimeq",
    "\\subseteqq",
    "\\Subset",
    "\\preccurlyeq",
    "\\curlyeqprec",
    "\\precsim",
    "\\precapprox",
    "\\vartriangleleft",
    "\\Vvdash",
    "\\bumpeq",
    "\\Bumpeq",
    "\\geqq",
    "\\geqslant",
    "\\eqslantgtr",
    "\\gtrsim",
    "\\gtrapprox",
    "\\eqsim",
    "\\gtrdot",
    "\\ggg",
    "\\gtrless",
    "\\gtreqless",
    "\\gtreqqless",
    "\\eqcirc",
    "\\circeq",
    "\\triangleq",
    "\\thicksim",
    "\\thickapprox",
    "\\supseteqq",
    "\\Supset",
    "\\succcurlyeq",
    "\\curlyeqsucc",
    "\\succsim",
    "\\succapprox",
    "\\vartriangleright",
    "\\shortmid",
    "\\shortparallel",
    "\\between",
    "\\pitchfork",
    "\\varpropto",
    "\\blacktriangleleft",
    "\\therefore",
    "\\backepsilon",
    "\\blacktriangleright",
    "\\because",
    "\\nleqslant",
    "\\nleqq",
    "\\lneq",
    "\\lneqq",
    "\\lvertneqq",
    "\\lnsim",
    "\\lnapprox",
    "\\nprec",
    "\\npreceq",
    "\\precneqq",
    "\\precnsim",
    "\\precnapprox",
    "\\nsim",
    "\\nshortmid",
    "\\nvdash",
    "\\nVdash",
    "\\ntriangleleft",
    "\\ntrianglelefteq",
    "\\nsubseteq",
    "\\nsubseteqq",
    "\\varsubsetneq",
    "\\subsetneqq",
    "\\varsubsetneqq",
    "\\ngtr",
    "\\ngeqslant",
    "\\ngeqq",
    "\\gneq",
    "\\gneqq",
    "\\gvertneqq",
    "\\gnsim",
    "\\gnapprox",
    "\\nsucc",
    "\\nsucceq",
    "\\succneqq",
    "\\succnsim",
    "\\succnapprox",
    "\\ncong",
    "\\nshortparallel",
    "\\nparallel",
    "\\nvDash",
    "\\nVDash",
    "\\ntriangleright",
    "\\ntrianglerighteq",
    "\\nsupseteq",
    "\\nsupseteqq",
    "\\varsupsetneq",
    "\\supsetneqq",
    "\\varsupsetneqq",
    "\\leftleftarrows",
    "\\leftrightarrows",
    "\\Lleftarrow",
    "\\leftarrowtail",
    "\\looparrowleft",
    "\\leftrightharpoons",
    "\\curvearrowleft",
    "\\circlearrowleft",
    "\\Lsh",
    "\\upuparrows",
    "\\rightrightarrows",
    "\\rightleftarrows",
    "\\Rrightarrow",
    "\\rightarrowtail",
    "\\looparrowright",
    "\\curvearrowright",
    "\\circlearrowright",
    "\\Rsh",
    "\\downdownarrows",
    "\\multimap",
    "\\leftrightsquigarrow",
    "\\rightsquigarrow",
    "\\nLeftarrow",
    "\\nleftrightarrow",
    "\\nRightarrow",
    "\\nLeftrightarrow",

    //"\\mathit", // in texvc, but ams not actually required
    //"\\mathrm", // in texvc, but ams not actually required
    //"\\mathord", // in texvc, but ams not actually required
    //"\\mathop", // in texvc, but ams not actually required
    //"\\mathbin", // in texvc, but ams not actually required
    //"\\mathrel", // in texvc, but ams not actually required
    //"\\mathopen", // in texvc, but ams not actually required
    //"\\mathclose", // in texvc, but ams not actually required
    //"\\mathpunct", // in texvc, but ams not actually required
    "\\boldsymbol",
    "\\mathbb",
    //"\\mathbf", // in texvc, but ams not actually required
    //"\\mathsf", // in texvc, but ams not actually required
    //"\\mathcal", // in texvc, but ams not actually required
    //"\\mathtt", // in texvc, but ams not actually required
    "\\mathfrak",
    "\\operatorname",
	"\\mathbb{R}"
]);

module.exports.cancel_required = arr2set([
    "\\bcancel",
    "\\cancel",
    "\\xcancel",
    "\\cancelto"
]);

module.exports.color_required = arr2set([
    "\\color",
    "\\pagecolor",
    "\\definecolor"
]);

module.exports.euro_required = arr2set([
    "\\euro",
    "\\geneuro",
    "\\geneuronarrow",
    "\\geneurowide",
    "\\officialeuro"
]);

module.exports.teubner_required = arr2set([
    "\\Coppa",
    "\\coppa",
    "\\Digamma",
    "\\Koppa",
    "\\koppa",
    "\\Sampi",
    "\\sampi",
    "\\Stigma",
    "\\stigma",
    "\\varstigma"
]);
